ci/supervisor/config/config.go

```diff
@@ -0,0 +1,329 @@
+	MinorReleasePeriodicity string `yaml:"minor-release-periodicity"`
+	MinorReleaseArg         string `yaml:"minor-release-arg"`
+}
+
+// SupervisorDefaults all the default configuration for the repository tasks
+type SupervisorDefaults struct {
+	Settings       DefaultSettings
+	ProwPlugins    map[string]ProwPlugin
+	CodeCoverage   CodeCoverageSettings
+	NightlyRelease NightlyReleaseSettings
+	AutoRelease    AutoReleaseSettings
+}
+
+// SupervisorConfig holds all the configuration for supervisor
+type SupervisorConfig struct {
+	Defaults     SupervisorDefaults
+	Repositories map[string][]string
+}
+
+// parsePlugins unmarshals a Prow plugin section into the given map
+func parsePlugins(pluginMap map[string]ProwPlugin, rawData yamlMap) error {
+	for p := range rawData {
+		pluginName, _ := p.(string)
+		plugin := ProwPlugin{Enabled: true}
+		badSetting := fmt.Errorf("valid setting for plugin %q is one of [disabled|no|false|<configuration>]", pluginName)
+		// Setting is a simple enable/disable string
+		if v, ok := rawData[pluginName].(string); ok {
+			if v == "disabled" || v == "no" || v == "false" {
+				plugin.Enabled = false
+			} else if v == "enabled" || v == "yes" || v == "true" {
+				plugin.Enabled = true
+			} else {
+				return badSetting
+			}
+			// Otherwise it's a raw plugin setting
+		} else {
+			if p, ok := rawData[pluginName].(yamlMap); ok {
+				plugin.Parameters = p
+			} else {
+				return badSetting
+			}
+		}
+		pluginMap[pluginName] = plugin
+	}
+	return nil
+}
+
+// UnmarshalYAML unmarshals a Defaults struct, correctly unmarshalling the Prow plugin section
+func (d *Defaults) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	// Expected fields
+	var structuredData struct {
+		Settings     DefaultSettings
+		ProwPlugins  yamlMap              `yaml:"prow-plugins"`
+		CodeCoverage CodeCoverageSettings `yaml:"code-coverage"`
+	}
+	if err := unmarshal(&structuredData); err != nil {
+		return err
+	}
+	d.Settings = structuredData.Settings
+	d.CodeCoverage = structuredData.CodeCoverage
+	d.ProwPlugins = make(map[string]ProwPlugin, 0)
+	return parsePlugins(d.ProwPlugins, structuredData.ProwPlugins)
+}
+
+// ParseRepoConfig unmarshals a repository config YAML
+func ParseRepoConfig(content []byte) (RepoConfig, error) {
+	r := RepoConfig{}
+	r.PresubmitTasks = make(map[string]PresubmitTask, 0)
+	r.PeriodicTasks = make(map[string]PeriodicTask, 0)
+	r.PostsubmitTasks = make(map[string]PostsubmitTask, 0)
+
+	// Extract the raw YAML
+	var rawConfig map[string]interface{}
+	if err := yaml.Unmarshal(content, &rawConfig); err != nil {
+		return RepoConfig{}, err
+	}
+
+	// Unmarshal the defaults section
+	if d, err := yaml.Marshal(rawConfig["defaults"]); err == nil {
+		if err := yaml.Unmarshal(d, &r.Defaults); err != nil {
+			return RepoConfig{}, err
+		}
+	} else {
+		return RepoConfig{}, err
+	}
+
+	// Unmarshal the presubmit-tasks section
+	if value, ok := rawConfig["presubmit-tasks"].(yamlMap); ok {
```

Comment:
might just be my personal preference, I feel like this block and following 2 blocks can be something like this:
```
r.LoadPresubmitTasksFromRawConfigOrFail(rawConfig)
r.LoadPeriodicTasksFromRawConfigOrFail(rawConfig)
r.LoadPostsubmitTasksFromRawConfigOrFail(rawConfig)
```